pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,327 @@
+	if err := w.ofClient.InstallMulticastFlow(e.group); err != nil {
+		return err
+	}
+	status := &GroupMemberStatus{
+		group:        e.group,
+		lastProbe:    e.time,
+		localMembers: sets.NewString(e.iface.InterfaceName),
+	}
+	// TODO: add multicast routing entries for the new added multicast group before add status into groupCache.
+	w.groupCache.Add(status)
+	klog.InfoS("Added new multicast group to cache", "group", e.group, "interface", e.iface.InterfaceName)
+	return nil
+}
+
+// updateGroupMemberStatus updates the group status in groupCache. If a "join" message is sent from an existing member,
+// only updates the lastProbe time. If a "join" message is sent from an "unknown" member, updates the lastProbe time and
+// adds the new member into the group's local member set. If a "leave" message is sent from an existing member, remove it from the
+// group's local member set, and if the member is the last one in local cache, a query message on the group is sent out
+// to check if there are still local members in the group.
+func (w *worker) updateGroupMemberStatus(obj interface{}, e *mcastGroupEvent) error {
+	status := obj.(*GroupMemberStatus)
+	status.mutex.Lock()
+	defer status.mutex.Unlock()
+	exist := status.localMembers.Has(e.iface.InterfaceName)
+	switch e.eType {
+	case groupJoin:
+		status.lastProbe = e.time
+		if !exist {
+			// TODO: process the new member in the multicast routing entries if necessary.
+			status.localMembers.Insert(e.iface.InterfaceName)
+			klog.InfoS("Added interface to Multicast group", "group", e.group.String(), "member", e.iface.InterfaceName)
+		}
+		w.groupCache.Update(status)
+	case groupLeave:
+		if exist {
+			// TODO: process the member in the multicast routing entries if necessary.
+			status.localMembers.Delete(e.iface.InterfaceName)
+			klog.InfoS("Member left the Multicast group", "group", e.group.String(), "member", e.iface.InterfaceName)
+			if len(status.localMembers) == 0 {
+				klog.InfoS("Check last member in multicast group", "group", e.group.String(), "member", e.iface.InterfaceName)
+				go w.checkLastMember(e.group)
+			}
+		}
+	}
+	return nil
+}
+
+// checkLastMember sends out a query message on the group to check if there are still members in the group. If no new
+// membership report is received in the max response time, the group is removed from groupCache.
+func (w *worker) checkLastMember(group net.IP) {
+	err := w.igmpSnooper.queryIGMP(group, queryVersion)
+	if err != nil {
+		klog.ErrorS(err, "Failed to seng IGMP query message", "group", group.String())
+		return
+	}
+	select {
+	case <-time.After(igmpMaxResponseTime):
+		klog.V(2).InfoS("Time up to wait for new membership report", "group", group.String())
+		// Notify worker to remove the stale group.
+		w.staleGroupCh <- group
+	case <-w.ctx.Done():
+		// End up the goroutine to avoid unnecessary leak when Agent is stopped.
+		klog.Info("Stop waiting for the response")
+		return
+	}
```

Comment:
Trying to understand what's the condition for the case that there is still valid members in the group. `queryIGMP` doesn't return any members and the group will be added to `staleGroupCh` anyway. Does it rely on `processGroupEvent` to check members before actually removing the group? If yes, the comment is misleading.

Reply:
Yes, it relies on `processGroupEvent` to check members before actually removing the group because there might be a race case that `checkLastMember` thinks a group is stale, but `checkLastMember` receives a new group membership report before handling the event from channel. Maybe I should revise the comments as "Notify worker that time is up waiting for new group member reports". what do you think?